extensions/amp-apester-media/0.1/monetization/companion/video.js

```diff
@@ -0,0 +1,151 @@
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {Services} from '../../../../../src/services';
+
+/**
+ * @param {!JsonObject} media
+ * @param {AmpElement} apesterElement
+ * @param {!JsonObject} consentObj
+ * @return {Promise}
+ */
+export function handleCompanionVideo(media, apesterElement, consentObj) {
+  const monetizationSettings = media['campaignData'] || {};
+  const companionCampaignOptions =
+    monetizationSettings['companionCampaignOptions'] || {};
+  const companionRawSettings = monetizationSettings['companionOptions'] || {};
+  if (
+    companionRawSettings.video &&
+    companionRawSettings.video.enabled &&
+    companionRawSettings.video.provider === 'sr'
+  ) {
+    const position = getCompanionPosition(companionRawSettings.video);
+    if (position && position !== 'floating') {
+      const macros = getSrMacros(
+        media,
+        companionCampaignOptions['companionCampaignId'],
+        apesterElement,
+        consentObj
+      );
+
+      constructCompanionSrElement(
+        companionRawSettings['video']['videoTag'],
+        position,
+        /** @type {!JsonObject} */ (macros),
+        apesterElement
+      );
+    }
+  }
+  return Promise.resolve();
+}
+/**
+ * @param {!JsonObject} video
+ * @return {string}
+ */
+function getCompanionPosition(video) {
+  if (video['companion']['enabled']) {
+    return 'above';
+  } else if (video['companion_below']['enabled']) {
+    return 'below';
+  } else if (video['floating']['enabled']) {
+    return 'floating';
+  }
+  return '';
+}
+
+/**
+ * @param {string} videoTag
+ * @param {string} position
+ * @param {!JsonObject} macros
+ * @param {AmpElement} apesterElement
+ */
+function constructCompanionSrElement(
+  videoTag,
+  position,
+  macros,
+  apesterElement
+) {
+  const size = getCompanionVideoAdSize(apesterElement);
+  const ampAd = apesterElement.ownerDocument.createElement('amp-ad');
+  ampAd.setAttribute('type', 'blade');
+  ampAd.setAttribute('data-blade_player_type', 'bladex');
+  ampAd.setAttribute('servingDomain', 'ssr.streamrail.net');
+  ampAd.setAttribute('width', size.width);
+  ampAd.setAttribute('height', size.height);
+  ampAd.setAttribute('data-blade_macros', JSON.stringify(macros));
+  ampAd.setAttribute('data-blade_player_id', videoTag);
+  ampAd.setAttribute('data-blade_api_key', '5857d2ee263dc90002000001');
+  ampAd.classList.add('amp-apester-companion');
+
+  const relativeElement =
+    position === 'below' ? apesterElement.nextSibling : apesterElement;
+
+  apesterElement.parentNode.insertBefore(ampAd, relativeElement);
```

Comment:
This pattern of going outside of your container and modifying it seems a bit weird and breaks encapsulation rules. It also seems like it could break AMP's static sizing constraints by causing content jumping. I'm not sure this is something we are okay with.

Is there a reason why this will not cause content jumping? How can we be sure that this is safe to do?

Reply:
Hi @sparhami ,

The goal here is to add the ability of inserting advertisements below and/or above our Apester unit. 

Since we're using amp extensions based amp-ad,
by following the amp-ad docs, (No ad available section https://amp.dev/documentation/components/amp-ad/),
We felt like amp-ad will take care about content jumping issues. Did we miss anything? 

About the encapsulation, we need our unit and the ad to be in a siblings hierarchy, do you have any suggestion regarding keeping the desired hierarchy without breaking encapsulation?  

Thanks, 

Almog.

Reply:
I'm not familiar enough with amp-ad to know if this will avoid content jumping.

@lannka Does this sound reasonable? I think we may want to do an i2i here and have a design review to iron out some questions.

Reply:
Inserting `amp-ad` this way can cause content jumping. We should consider first try to resize the slot through `attemptChangeHeight` (which can fail) then insert the amp-ad element.

Reply:
Just realized that `amp-ad` is inserted as a sibling of `amp-asper-media`. So we will need to insert a 0 height `amp-ad` first, and then call `attemptChangeHeight` to request a resize.

Re: encapsulation. I don't immediately see any potential breakage it can cause at this moment, but I agree that allowing extension modifying content outside it in this way makes me a bit nervous. Do we have any other extensions have the similar behavior? Do people foresee any issues? @ampproject/wg-runtime  @ampproject/wg-ui-and-a11y

Reply:
AMP defers the instantiation of a zero-height AMP element. that said, no ad is being requested until it gets resized. that's actually how `amp-auto-ads` injects `amp-ad` dynamically on the page.

could you try?

speaking of `amp-auto-ads`, one way to avoid the "encapsulation" issue above is to move the ads insertion function out of your `amp-asper-media` component to `amp-auto-ads`. That means your customer will use `amp-asper-media` as it is to load content, and use `amp-auto-ads` to dynamically place ads to the page.

Reply:
No, unfortunately this isn't possible, the publishers managed their ads configuration by Apester, they don't want to do any programming after the first initialisation, same use case they have on their regular (non AMP) sites.

And about the height issue, seems like attemptChangeSize in base-element has been called when we tried to load the video ad.

In the local test we have made, if there is no ad (no impression),the ad element isn't appear on the screen by following the amp-ad docs, (No ad available section https://amp.dev/documentation/components/amp-ad/).

Reply:
> No, unfortunately this isn't possible, the publishers manage the ads configuration by Apester, they don't want to do any programming after the first initialisation, same use case they have on their regular (non AMP) sites.

I suppose this is responding to the amp-auto-ads proposal? No programming needed from publisher. You will implement your ads feature in [amp-auto-ads](https://github.com/ampproject/amphtml/blob/master/extensions/amp-auto-ads/amp-auto-ads.md), and pub will tag there page like this:

```html
<amp-auto-ads type=apester>
</amp-auto-ads>

...

<amp-apester-media>
...
</amp-apester-media>
```

> And about the height issue, seems like attemptChangeSize in base-element has been invoked when we tried to load the video ad.

Did you insert `amp-ad` with height 0 and then call attemptChangeSize? Can you update the code to show how you did it?

Reply:
> > No, unfortunately this isn't possible, the publishers manage the ads configuration by Apester, they don't want to do any programming after the first initialisation, same use case they have on their regular (non AMP) sites.
> 
> I suppose this is responding to the amp-auto-ads proposal? No programming needed from publisher. You will implement your ads feature in [amp-auto-ads](https://github.com/ampproject/amphtml/blob/master/extensions/amp-auto-ads/amp-auto-ads.md), and pub will tag there page like this:
> 
> ```
> <amp-auto-ads type=apester>
> </amp-auto-ads>
> 
> ...
> 
> <amp-apester-media>
> ...
> </amp-apester-media>

Hi @lannka, 
From what I understood in that case, we'll to provide an ad element as well and this is not the service we provide to publishers, We'll need to repeat on some of the api requests we already did in amp-apester-element to get the data, and it's seems like at the end the result is the same. 

Is it possible to get guidance directly from you on slack or scheduled a meeting? We have many publishers waiting for that feature, from the Dumont group on Germany (pc-magazine.de, express.de, mopo.de etc), to TI media group and ESI media group, I think that a short meeting like that could take it further quickly.

Reply:
@almoghaimo we can chat in Slack, you can `@` me in the `#wg-ads` channel.